image/manifest.go

```diff
@@ -26,28 +26,297 @@ import (
+		SnapName: snapName,
+		Revision: revision,
+	}
+	return nil
+}
+
+// MarkValidationSetSeeded marks a validation-set as seeded. It verifies against any previously
+// set rules by SetAllowedValidationSet, and sets up new rules based on the snaps defined in the
+// validation set.
+// This relies on validation-set assertions being marked here the moment they are fetched by
+// the seedwriter, which should be done before the first call to MarkSnapRevisionSeeded.
+func (sm *SeedManifest) MarkValidationSetSeeded(vsa *asserts.ValidationSet, pinned bool) error {
+	vs := newSeedManifestValidationSet(vsa, pinned)
+	if _, ok := sm.vsSeeded[vs.Unique()]; ok {
+		return fmt.Errorf("cannot mark validation set %q as seeded, it has already been marked as such",
+			vs.Unique())
+	}
 
-// ReadSeedManifest reads a seed.manifest generated by ubuntu-image, and returns
-// a map containing the snap names and their revisions.
-func ReadSeedManifest(manifestFile string) (map[string]snap.Revision, error) {
+	// Check against any pre-defined restrictions.
+	if allowed, ok := sm.vsAllowed[vs.Unique()]; ok {
+		if allowed.Sequence != vs.Sequence {
+			return fmt.Errorf("sequence of %q (%d) does not match the allowed sequence (%d)",
+				vs.Unique(), vs.Sequence, allowed.Sequence)
+		}
+		if allowed.Pinned != pinned {
+			return fmt.Errorf("pinning of %q (%t) does not match the allowed pinning (%t)",
+				vs.Unique(), pinned, allowed.Pinned)
+		}
+	}
+
+	for _, sn := range vsa.Snaps() {
+		// Record only snaps that have a presence set, and a revision specified
+		if sn.Presence == asserts.PresenceInvalid {
+			continue
+		}
+		if sn.Revision <= 0 {
+			continue
+		}
+
+		// Update allowed snaps based on the validation-set.
+		if err := sm.SetAllowedSnapRevision(sn.SnapName(), snap.R(sn.Revision)); err != nil {
+			return err
+		}
+
+		// For book-keeping purposes we add the snap to the list of controlled
+		// snap revisions by the validation set.
+		vs.Snaps = append(vs.Snaps, sn.SnapName())
+	}
+
+	sm.vsSeeded[vs.Unique()] = vs
+	return nil
+}
+
+// AllowedRevision retrieves any specified revision rule for the snap
+// name.
+func (sm *SeedManifest) AllowedSnapRevision(snapName string) snap.Revision {
+	// TODO: Check seeded validation-sets as well.
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		return rev.Revision
+	}
+	return snap.Revision{}
+}
+
+// ValidationSetsAllowed returns the validation sets specified as allowed.
+func (sm *SeedManifest) AllowedValidationSets() []*SeedManifestValidationSet {
+	var vss []*SeedManifestValidationSet
+	for _, vs := range sm.vsAllowed {
+		vss = append(vss, vs)
+	}
+
+	// Sort for test consistency
+	sort.Slice(vss, func(i, j int) bool {
+		return vss[i].Unique() < vss[j].Unique()
+	})
+	return vss
+}
+
+func parsePinnedValidationSet(sm *SeedManifest, vs string) error {
+	acc, name, seq, err := snapasserts.ParseValidationSet(vs)
+	if err != nil {
+		return err
+	}
+	return sm.SetAllowedValidationSet(acc, name, seq, true)
```

Comment:
Here I don't like the bool for pined/unpined, I wonder if a `struct ValidationSetOptions{ seq, pined}` would be better here? Could be a followup to make your life easier of course :)

Reply:
Definitely could make it more readable, though I might do this in a followup at this point, but the suggestion is really good!